compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -697,19 +691,14 @@ impl<'tcx> TypeVisitor<'tcx> for OrphanChecker<'tcx> {
                 }
             }
             ty::Error(_) => ControlFlow::Break(OrphanCheckEarlyExit::LocalTy(ty)),
-            ty::Closure(..) | ty::Generator(..) | ty::GeneratorWitness(..) => {
-                if !self.tcx.features().specialization {
-                    // With specialization, there could be a blanket impl `impl<T: Send> Trait for T`
-                    // as well as an `impl Trait for OpaqueType`, where the hidden type of `OpaqueType`
-                    // is a closure or generator. See tests/ui/type-alias-impl-trait/issue-104817.rs for
-                    // an example.
-                    self.tcx.sess.delay_span_bug(
-                        DUMMY_SP,
-                        format!("ty_is_local invoked on closure or generator: {:?}", ty),
-                    );
+            ty::Closure(did, ..) | ty::Generator(did, ..) => {
+                if self.def_id_is_local(did) {
+                    ControlFlow::Break(OrphanCheckEarlyExit::LocalTy(ty))
+                } else {
+                    self.found_non_local_ty(ty)
                 }
-                ControlFlow::Break(OrphanCheckEarlyExit::LocalTy(ty))
             }
+            ty::GeneratorWitness(tys) => tys.visit_with(self),
```

Comment:
this is for the impl
```rust
impl<T, U> AutoTrait for Witness<T, U>
where
    FieldTy: AutoTrait,
    FieldTy: AutoTrait,
```
this should only be observable in evaluate itself when checking whether we have to consider some trait goal unknowable. This impl will always be the only impl that exists so you can break with `LocalTy` here.

It would be nice to assert that this can only be reached for auto traits.

please update pr name + descr, then r=me